File: HappyCollisions/Physics/WorldPhysics.cs
using HappyCollisions.Actors;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace HappyCollisions.Physics
{
    class WorldPhysics
    {
        private static readonly double G_CONSTANT = 10.0f;

        private readonly List<IActor> actors = new List<IActor>();

        public void Tick(double dt = 0.01)
        {
            MoveActors(dt);
            ApplyGlobalAcceleration(dt);
        }

        public void AddActor(IActor actor)
        {
            this.actors.Add(actor);
        }

        public List<IActor> Actors { get => actors; }

        private void ApplyGlobalAcceleration(double dt)
        {
            actors.ForEach(actor => actor.VY += G_CONSTANT * dt);
        }

        private void MoveActors(double dt)
        {
            actors.ForEach(actor => { actor.X += actor.VX * dt; actor.Y += actor.VY * dt; });
        }
    }
}

File: CollisionFun/Physics/WorldPhysics.fs
module Physics.WorldPhysics

open OpenTK
open Interfaces
open Actors.Operations

type CustomWorldPhysics () =
    let mutable actors : Actor list = []
    let GRAVITY_CONSTANT : Vector2d = Vector2d(0.0, -0.01)
    let moveActors (delta : float) : unit =
        actors
        |> List.map ActorsData
        |> List.map (fun data -> data.Move (data.Velocity * delta))
        |> ignore
    let applyGravity (delta : float) : unit =
        actors
        |> List.map (AccelerateActor (GRAVITY_CONSTANT * delta))
        |> ignore
    interface IWorldPhysics with
        member __.AddActor (actor : Actor): unit = 
            actors <- actor::actors
        member __.RemoveActor (actor: Actor): unit = 
            actors <- actors
                      |> List.filter (fun a -> a <> actor)
        member __.Tick(delta: float): unit = 
            applyGravity delta
            moveActors delta
        member __.Actors : Actor list =
            actors